File: admin/src/types/marketplace.ts
/**
 * Types for Unified Marketplace Module
 * Phase 5: Omnichannel Integration and Management
 */

// ============================================================================
// MARKETPLACE INTEGRATIONS
// ============================================================================

export type MarketplaceType =
  | 'mercado_livre'
  | 'amazon'
  | 'shopee'
  | 'magalu'
  | 'b2w'
  | 'tiktok_shop'
  | 'shopify'
  | 'woocommerce'
  | 'custom';

export type SyncStatus = 'success' | 'error' | 'partial' | 'pending';
export type ListingStatus = 'active' | 'paused' | 'out_of_stock' | 'error';
export type UnifiedOrderStatus = 'pending' | 'processing' | 'shipped' | 'delivered' | 'cancelled';
export type PaymentStatus = 'pending' | 'approved' | 'rejected';

export interface MarketplaceIntegration {
  id: string;
  workspace_id: number;
  marketplace: MarketplaceType;
  name: string;

  // Credentials
  credentials: {
    client_id?: string;
    client_secret?: string;
    access_token?: string;
    refresh_token?: string;
    store_url?: string;
    api_key?: string;
    seller_id?: string;
  };

  // Configuration
  is_active: boolean;
  auto_sync: boolean;
  sync_frequency: number; // minutes

  // Mapping
  default_warehouse_id: string;
  price_adjustment: {
    type: 'percentage' | 'fixed';
    value: number;
  };

  // Sync settings
  sync_products: boolean;
  sync_orders: boolean;
  sync_stock: boolean;

  // Last sync
  last_sync_at?: Date;
  last_sync_status: SyncStatus;
  last_sync_error?: string;
  last_sync_summary?: {
    products_synced: number;
    orders_imported: number;
    stock_updated: number;
    errors: number;
  };

  // Statistics
  total_listings: number;
  active_listings: number;
  total_orders: number;
  total_revenue: number;

  created_at: Date;
  updated_at: Date;
}

export interface ProductListing {
  id: string;
  product_id: string;
  product_name: string;
  product_sku: string;
  marketplace_integration_id: string;
  marketplace: MarketplaceType;

  // External IDs
  external_id: string;
  external_sku: string;
  listing_url: string;

  // Price and stock
  price: number;
  original_price?: number;
  stock_quantity: number;

  // Status
  status: ListingStatus;
  is_synced: boolean;
  sync_enabled: boolean;

  // Customization
  title?: string; // Custom title
  description?: string; // Custom description
  images?: string[]; // Image URLs
  category?: string;
  tags?: string[];

  // Metrics
  views: number;
  sales: number;
  conversion_rate: number;
  revenue: number;

  // Shipping
  free_shipping: boolean;
  shipping_cost?: number;

  // Sync
  last_synced_at?: Date;
  sync_errors?: string[];
  pending_changes: boolean;

  created_at: Date;
  updated_at: Date;
}

export interface UnifiedOrder {
  id: string;
  workspace_id: number;

  // Origin
  marketplace_integration_id: string;
  marketplace: MarketplaceType;
  marketplace_name: string;
  external_order_id: string;
  external_order_number: string;

  // Customer data (normalized)
  customer: {
    name: string;
    email: string;
    phone: string;
    document: string;
    external_customer_id?: string;
  };

  // Items
  items: UnifiedOrderItem[];

  // Shipping
  shipping: {
    method: string;
    cost: number;
    address: ShippingAddress;
    tracking_code?: string;
    estimated_delivery?: Date;
  };

  // Payment
  payment: {
    method: string;
    status: PaymentStatus;
    paid_at?: Date;
    transaction_id?: string;
  };

  // Totals
  subtotal: number;
  shipping_cost: number;
  discount: number;
  tax: number;
  total: number;

  // Unified status
  status: UnifiedOrderStatus;

  // Internal mapping
  sale_id?: string; // ID da venda interna criada
  processed: boolean;
  processing_errors?: string[];

  // Import
  imported_at: Date;
  imported_by?: string;

  // Dates
  order_date: Date;
  created_at: Date;
  updated_at: Date;
}

export interface UnifiedOrderItem {
  id: string;
  product_id?: string; // Internal product ID
  external_product_id: string;
  product_name: string;
  sku: string;
  quantity: number;
  unit_price: number;
  discount: number;
  tax: number;
  total: number;

  // Listing reference
  listing_id?: string;
}

export interface ShippingAddress {
  street: string;
  number: string;
  complement?: string;
  neighborhood: string;
  city: string;
  state: string;
  postal_code: string;
  country: string;
  reference?: string;
}

// ============================================================================
// SYNC STRATEGIES
// ============================================================================

export type ProductSelectionType = 'all' | 'category' | 'custom' | 'tag';
export type PriceStrategyType = 'same' | 'markup' | 'markdown' | 'custom';
export type ConflictResolutionType = 'manual' | 'system_wins' | 'marketplace_wins' | 'average';

export interface SyncStrategy {
  id: string;
  name: string;
  workspace_id: number;

  // Product selection
  product_selection: ProductSelectionType;
  category_ids?: string[];
  product_ids?: string[];
  tags?: string[];

  // Rules
  rules: {
    // Stock
    minimum_stock_to_publish: number;
    reserve_stock_percentage: number; // % to reserve
    auto_pause_on_low_stock: boolean;
    low_stock_threshold: number;

    // Price
    price_strategy: PriceStrategyType;
    markup_percentage?: number;
    markdown_percentage?: number;
    custom_prices?: Record<string, number>; // product_id -> price
    round_prices: boolean;
    round_to?: number; // e.g., 0.99

    // Content
    use_custom_title: boolean;
    title_template?: string; // e.g., "{product_name} - {brand}"
    use_custom_description: boolean;
    append_to_description?: string;

    // Images
    optimize_images: boolean;
    max_images: number;
  };

  // Target marketplaces
  target_marketplaces: string[];

  // Schedule
  auto_sync_enabled: boolean;
  sync_frequency: number; // minutes

  is_active: boolean;
  created_at: Date;
  updated_at: Date;
}

export interface SyncConflict {
  id: string;
  type: 'stock_discrepancy' | 'price_mismatch' | 'product_not_found' | 'duplicate_listing' | 'data_conflict';

  // Details
  product_id: string;
  product_name: string;
  listing_id?: string;
  marketplace_integration_id: string;
  marketplace: MarketplaceType;

  // Values
  system_value: any;
  marketplace_value: any;
  field_name: string;

  // Resolution
  resolution_strategy: ConflictResolutionType;
  resolved: boolean;
  resolved_at?: Date;
  resolved_by?: string;
  resolution_notes?: string;

  // Severity
  severity: 'low' | 'medium' | 'high' | 'critical';
  auto_resolvable: boolean;

  created_at: Date;
}

export interface SyncJob {
  id: string;
  workspace_id: number;
  marketplace_integration_id: string;
  marketplace: MarketplaceType;

  // Job details
  type: 'full_sync' | 'incremental' | 'stock_only' | 'prices_only' | 'orders_only';
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

  // Progress
  total_items: number;
  processed_items: number;
  successful_items: number;
  failed_items: number;
  progress_percentage: number;

  // Results
  result: {
    products_created: number;
    products_updated: number;
    listings_created: number;
    listings_updated: number;
    orders_imported: number;
    stock_synced: number;
    errors: SyncError[];
  };

  // Timing
  started_at?: Date;
  completed_at?: Date;
  duration_seconds?: number;

  // Error handling
  error_message?: string;
  retry_count: number;
  max_retries: number;

  created_at: Date;
  updated_at: Date;
}

export interface SyncError {
  item_id: string;
  item_type: 'product' | 'listing' | 'order' | 'stock';
  error_code: string;
  error_message: string;
  retry_possible: boolean;
}

// ============================================================================
// MARKETPLACE ANALYTICS
// ============================================================================

export interface MarketplacePerformance {
  marketplace_integration_id: string;
  marketplace: MarketplaceType;
  period_start: Date;
  period_end: Date;

  // Sales metrics
  total_orders: number;
  total_revenue: number;
  avg_order_value: number;
  items_sold: number;

  // Growth
  revenue_growth: number; // % vs previous period
  orders_growth: number;

  // Listings
  total_listings: number;
  active_listings: number;
  out_of_stock_listings: number;

  // Performance
  conversion_rate: number; // %
  avg_views_per_listing: number;
  avg_sales_per_listing: number;

  // Issues
  sync_errors: number;
  conflicts: number;
  failed_orders: number;

  // Top products
  top_products: {
    product_id: string;
    product_name: string;
    quantity_sold: number;
    revenue: number;
  }[];
}

export interface ChannelComparison {
  period_start: Date;
  period_end: Date;

  channels: {
    marketplace: MarketplaceType;
    name: string;

    // Sales
    orders: number;
    revenue: number;
    revenue_percentage: number;

    // Efficiency
    avg_order_value: number;
    conversion_rate: number;

    // Costs
    fees: number;
    net_revenue: number;
    profit_margin: number;
  }[];

  // Overall
  total_revenue: number;
  total_orders: number;
  best_performing_channel: MarketplaceType;
  worst_performing_channel: MarketplaceType;
}

// ============================================================================
// CATALOG MANAGEMENT
// ============================================================================

export interface BulkPublishJob {
  id: string;
  workspace_id: number;
  name: string;

  // Products
  product_ids: string[];
  total_products: number;

  // Target marketplaces
  target_marketplaces: string[];

  // Settings
  sync_strategy_id?: string;
  price_adjustment?: {
    type: 'percentage' | 'fixed';
    value: number;
  };

  // Status
  status: 'pending' | 'processing' | 'completed' | 'failed';
  progress: {
    total: number;
    published: number;
    failed: number;
    percentage: number;
  };

  // Results
  results: {
    marketplace: MarketplaceType;
    successful: number;
    failed: number;
    errors: string[];
  }[];

  started_at?: Date;
  completed_at?: Date;
  created_at: Date;
}

export interface PriceRule {
  id: string;
  workspace_id: number;
  name: string;

  // Application
  applies_to: 'all' | 'category' | 'products' | 'marketplace';
  category_ids?: string[];
  product_ids?: string[];
  marketplace_ids?: string[];

  // Rule
  rule_type: 'markup' | 'markdown' | 'fixed' | 'dynamic';
  value: number;

  // Dynamic pricing
  dynamic_conditions?: {
    based_on: 'stock_level' | 'competition' | 'demand' | 'time';
    min_price?: number;
    max_price?: number;
    rules: {
      condition: string;
      adjustment: number;
    }[];
  };

  // Schedule
  start_date?: Date;
  end_date?: Date;
  active: boolean;

  // Priority
  priority: number; // Higher number = higher priority

  created_at: Date;
  updated_at: Date;
}

export interface StockReservation {
  id: string;
  product_id: string;
  marketplace_integration_id: string;
  marketplace: MarketplaceType;

  quantity_reserved: number;
  reason: 'pending_order' | 'manual' | 'safety_stock';

  expires_at?: Date;
  released: boolean;
  released_at?: Date;

  created_at: Date;
}

// ============================================================================
// DASHBOARD & WIDGETS
// ============================================================================

export interface MarketplaceDashboard {
  workspace_id: number;
  period: {
    start: Date;
    end: Date;
  };

  // Overview
  overview: {
    total_integrations: number;
    active_integrations: number;
    total_listings: number;
    active_listings: number;
    total_orders: number;
    total_revenue: number;
  };

  // Recent activity
  recent_orders: UnifiedOrder[];
  recent_syncs: SyncJob[];
  pending_conflicts: number;

  // Alerts
  alerts: {
    type: 'error' | 'warning' | 'info';
    message: string;
    marketplace?: MarketplaceType;
    action_required: boolean;
    created_at: Date;
  }[];

  // Quick stats
  stats: {
    orders_today: number;
    revenue_today: number;
    sync_errors_today: number;
    avg_sync_time: number; // seconds
  };
}

File: backend/app/models/marketplace.py
"""
Marketplace Integration Models
Multi-tenant marketplace integration for omnichannel management
"""
import enum
from datetime import datetime
from sqlalchemy import (
    Column, Integer, String, Float, Boolean, DateTime,
    ForeignKey, Text, JSON, Enum as SQLEnum, Index
)
from sqlalchemy.orm import relationship
from app.core.database import Base


# ============================================================================
# ENUMS
# ============================================================================

class MarketplaceType(str, enum.Enum):
    MERCADO_LIVRE = "mercado_livre"
    AMAZON = "amazon"
    SHOPEE = "shopee"
    MAGALU = "magalu"
    B2W = "b2w"
    TIKTOK_SHOP = "tiktok_shop"
    SHOPIFY = "shopify"
    WOOCOMMERCE = "woocommerce"
    CUSTOM = "custom"


class SyncStatus(str, enum.Enum):
    SUCCESS = "success"
    ERROR = "error"
    PARTIAL = "partial"
    PENDING = "pending"


class ListingStatus(str, enum.Enum):
    ACTIVE = "active"
    PAUSED = "paused"
    OUT_OF_STOCK = "out_of_stock"
    ERROR = "error"


class UnifiedOrderStatus(str, enum.Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    SHIPPED = "shipped"
    DELIVERED = "delivered"
    CANCELLED = "cancelled"


class PaymentStatus(str, enum.Enum):
    PENDING = "pending"
    APPROVED = "approved"
    REJECTED = "rejected"


class SyncJobStatus(str, enum.Enum):
    RUNNING = "running"
    COMPLETED = "completed"
    FAILED = "failed"


class SyncJobType(str, enum.Enum):
    FULL_SYNC = "full_sync"
    INCREMENTAL = "incremental"
    STOCK_ONLY = "stock_only"
    ORDERS_ONLY = "orders_only"


class ConflictType(str, enum.Enum):
    STOCK_DISCREPANCY = "stock_discrepancy"
    PRICE_MISMATCH = "price_mismatch"
    DATA_CONFLICT = "data_conflict"


class ConflictSeverity(str, enum.Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"
    CRITICAL = "critical"


# ============================================================================
# MODELS
# ============================================================================

class MarketplaceIntegration(Base):
    """
    Marketplace Integration
    Stores configuration for each marketplace connection
    """
    __tablename__ = "marketplace_integrations"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)

    marketplace = Column(SQLEnum(MarketplaceType), nullable=False, index=True)
    name = Column(String(255), nullable=False)

    # Credentials (stored encrypted in production)
    credentials = Column(JSON, nullable=False)  # client_id, client_secret, access_token, etc.

    # Configuration
    is_active = Column(Boolean, default=True, nullable=False)
    auto_sync = Column(Boolean, default=True, nullable=False)
    sync_frequency = Column(Integer, default=30, nullable=False)  # minutes

    # Mapping
    default_warehouse_id = Column(Integer, ForeignKey("warehouses.id"), nullable=True)
    price_adjustment_type = Column(String(50), nullable=True)  # 'percentage' or 'fixed'
    price_adjustment_value = Column(Float, default=0, nullable=False)

    # Sync settings
    sync_products = Column(Boolean, default=True, nullable=False)
    sync_orders = Column(Boolean, default=True, nullable=False)
    sync_stock = Column(Boolean, default=True, nullable=False)

    # Last sync
    last_sync_at = Column(DateTime, nullable=True)
    last_sync_status = Column(SQLEnum(SyncStatus), default=SyncStatus.PENDING, nullable=False)
    last_sync_error = Column(Text, nullable=True)
    last_sync_summary = Column(JSON, nullable=True)  # products_synced, orders_imported, etc.

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="marketplace_integrations")
    warehouse = relationship("Warehouse", foreign_keys=[default_warehouse_id])
    listings = relationship("ProductListing", back_populates="integration", cascade="all, delete-orphan")
    orders = relationship("UnifiedOrder", back_populates="integration", cascade="all, delete-orphan")
    sync_jobs = relationship("SyncJob", back_populates="integration", cascade="all, delete-orphan")
    conflicts = relationship("SyncConflict", back_populates="integration", cascade="all, delete-orphan")

    # Indexes
    __table_args__ = (
        Index('idx_marketplace_workspace', 'workspace_id', 'marketplace'),
        Index('idx_marketplace_active', 'is_active'),
    )


class ProductListing(Base):
    """
    Product Listing on Marketplace
    Maps internal products to marketplace listings
    """
    __tablename__ = "product_listings"

    id = Column(Integer, primary_key=True, index=True)
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False, index=True)
    marketplace_integration_id = Column(Integer, ForeignKey("marketplace_integrations.id"), nullable=False, index=True)

    # External IDs
    external_id = Column(String(255), nullable=False, index=True)
    external_sku = Column(String(255), nullable=True)
    listing_url = Column(Text, nullable=True)

    # Price and stock
    price = Column(Float, nullable=False)
    original_price = Column(Float, nullable=True)
    stock_quantity = Column(Integer, default=0, nullable=False)

    # Status
    status = Column(SQLEnum(ListingStatus), default=ListingStatus.ACTIVE, nullable=False, index=True)
    is_synced = Column(Boolean, default=False, nullable=False)
    sync_enabled = Column(Boolean, default=True, nullable=False)

    # Customization
    title = Column(String(500), nullable=True)  # Custom title
    description = Column(Text, nullable=True)  # Custom description
    images = Column(JSON, nullable=True)  # Array of image URLs
    category = Column(String(255), nullable=True)
    tags = Column(JSON, nullable=True)  # Array of tags

    # Metrics
    views = Column(Integer, default=0, nullable=False)
    sales = Column(Integer, default=0, nullable=False)
    conversion_rate = Column(Float, default=0, nullable=False)
    revenue = Column(Float, default=0, nullable=False)

    # Shipping
    free_shipping = Column(Boolean, default=False, nullable=False)
    shipping_cost = Column(Float, nullable=True)

    # Sync
    last_synced_at = Column(DateTime, nullable=True)
    sync_errors = Column(JSON, nullable=True)  # Array of error messages
    pending_changes = Column(Boolean, default=False, nullable=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    product = relationship("Product", back_populates="marketplace_listings")
    integration = relationship("MarketplaceIntegration", back_populates="listings")

    # Indexes
    __table_args__ = (
        Index('idx_listing_marketplace', 'marketplace_integration_id', 'status'),
        Index('idx_listing_product', 'product_id'),
        Index('idx_listing_external', 'external_id'),
    )


class UnifiedOrder(Base):
    """
    Unified Order from Marketplace
    Normalizes orders from different marketplaces
    """
    __tablename__ = "unified_orders"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)
    marketplace_integration_id = Column(Integer, ForeignKey("marketplace_integrations.id"), nullable=False, index=True)

    # External IDs
    external_order_id = Column(String(255), nullable=False, index=True)
    external_order_number = Column(String(255), nullable=False)

    # Customer data (normalized, stored as JSON)
    customer_data = Column(JSON, nullable=False)  # name, email, phone, document, etc.

    # Items (stored as JSON array)
    items = Column(JSON, nullable=False)

    # Shipping
    shipping_data = Column(JSON, nullable=False)  # method, cost, address, tracking, etc.

    # Payment
    payment_data = Column(JSON, nullable=False)  # method, status, paid_at, transaction_id

    # Totals
    subtotal = Column(Float, nullable=False)
    shipping_cost = Column(Float, default=0, nullable=False)
    discount = Column(Float, default=0, nullable=False)
    tax = Column(Float, default=0, nullable=False)
    total = Column(Float, nullable=False)

    # Unified status
    status = Column(SQLEnum(UnifiedOrderStatus), default=UnifiedOrderStatus.PENDING, nullable=False, index=True)

    # Internal mapping
    sale_id = Column(Integer, ForeignKey("sales.id"), nullable=True, index=True)
    processed = Column(Boolean, default=False, nullable=False)
    processing_errors = Column(JSON, nullable=True)

    # Import
    imported_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    imported_by = Column(String(255), nullable=True)

    # Dates
    order_date = Column(DateTime, nullable=False)
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="unified_orders")
    integration = relationship("MarketplaceIntegration", back_populates="orders")
    sale = relationship("Sale", foreign_keys=[sale_id])

    # Indexes
    __table_args__ = (
        Index('idx_unified_order_marketplace', 'marketplace_integration_id', 'status'),
        Index('idx_unified_order_external', 'external_order_id'),
        Index('idx_unified_order_date', 'order_date'),
        Index('idx_unified_order_processed', 'processed'),
    )


class SyncJob(Base):
    """
    Sync Job
    Tracks synchronization jobs with marketplaces
    """
    __tablename__ = "sync_jobs"

    id = Column(Integer, primary_key=True, index=True)
    workspace_id = Column(Integer, ForeignKey("workspaces.id"), nullable=False, index=True)
    marketplace_integration_id = Column(Integer, ForeignKey("marketplace_integrations.id"), nullable=False, index=True)

    type = Column(SQLEnum(SyncJobType), nullable=False)
    status = Column(SQLEnum(SyncJobStatus), default=SyncJobStatus.RUNNING, nullable=False, index=True)

    # Progress
    total_items = Column(Integer, default=0, nullable=False)
    processed_items = Column(Integer, default=0, nullable=False)
    successful_items = Column(Integer, default=0, nullable=False)
    failed_items = Column(Integer, default=0, nullable=False)
    progress_percentage = Column(Float, default=0, nullable=False)

    # Result summary
    result = Column(JSON, nullable=True)  # products_created, products_updated, etc.

    # Timing
    started_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    completed_at = Column(DateTime, nullable=True)
    duration_seconds = Column(Integer, nullable=True)

    # Retry
    retry_count = Column(Integer, default=0, nullable=False)
    max_retries = Column(Integer, default=3, nullable=False)

    # Error log
    error_log = Column(Text, nullable=True)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    workspace = relationship("Workspace", back_populates="sync_jobs")
    integration = relationship("MarketplaceIntegration", back_populates="sync_jobs")

    # Indexes
    __table_args__ = (
        Index('idx_sync_job_marketplace', 'marketplace_integration_id', 'status'),
        Index('idx_sync_job_created', 'created_at'),
    )


class SyncConflict(Base):
    """
    Sync Conflict
    Tracks conflicts between system and marketplace data
    """
    __tablename__ = "sync_conflicts"

    id = Column(Integer, primary_key=True, index=True)
    type = Column(SQLEnum(ConflictType), nullable=False, index=True)

    # Product reference
    product_id = Column(Integer, ForeignKey("products.id"), nullable=False, index=True)
    listing_id = Column(Integer, ForeignKey("product_listings.id"), nullable=True, index=True)
    marketplace_integration_id = Column(Integer, ForeignKey("marketplace_integrations.id"), nullable=False, index=True)

    # Conflict details
    field_name = Column(String(100), nullable=False)
    system_value = Column(String(500), nullable=False)
    marketplace_value = Column(String(500), nullable=False)

    # Resolution
    resolution_strategy = Column(String(50), default='manual', nullable=False)  # manual, system_wins, marketplace_wins
    resolved = Column(Boolean, default=False, nullable=False, index=True)
    resolved_at = Column(DateTime, nullable=True)
    resolved_by = Column(String(255), nullable=True)

    # Severity
    severity = Column(SQLEnum(ConflictSeverity), default=ConflictSeverity.MEDIUM, nullable=False)
    auto_resolvable = Column(Boolean, default=False, nullable=False)

    # Timestamps
    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)

    # Relationships
    product = relationship("Product")
    listing = relationship("ProductListing")
    integration = relationship("MarketplaceIntegration", back_populates="conflicts")

    # Indexes
    __table_args__ = (
        Index('idx_conflict_marketplace', 'marketplace_integration_id', 'resolved'),
        Index('idx_conflict_product', 'product_id'),
        Index('idx_conflict_severity', 'severity', 'resolved'),
    )
